### PR TITLE
[Execution][Block-STM] Create the final block-stm output in parallel

### PR DESCRIPTION
**Pull request by sitalkedia**
### Description

Profiling the execution code, we figured that the output creation which is currently being done in serial, consumes around 10% of the time for larger block size of ~20k. This change makes the output creation parallel.

### Test Plan
Existing UTs. 

Ran the single node performance benchmark on t2d-standard-60 machine and it gives us overall 4% TPS boost. The TPS goes from 31500 to 32800. 

Command - 

```
cargo run --profile performance -p aptos-executor-benchmark  -- --execution-threads 48 --generate-then-execute --transactions-per-sender 1 --block-size 25000 --split-ledger-db --use-sharded-state-merkle-db --skip-index-and-usage run-executor  --main-signer-accounts 100000  --data-dir ~/data/db2 --checkpoint-dir ~/data/chk --blocks 50
```

### CODE REVIEW

File: aptos-move/block-executor/src/txn_last_input_output.rs
Comment (gelash) [inline aptos-move/block-executor/src/txn_last_input_output.rs:352]: why do we need to change return type here?

File: aptos-move/block-executor/src/executor.rs
Comment (gelash) [inline aptos-move/block-executor/src/executor.rs:612]: I think this may be done more nicely as a part of commit_epilogue, maybe @ziaptos can see if that can be added here? 
https://github.com/aptos-labs/aptos-core/pull/10081

File: aptos-move/block-executor/src/executor.rs
Comment (gelash) [inline aptos-move/block-executor/src/executor.rs:633]: we can use Atomic type and have fetch min.

**Review by gelash — COMMENTED** [review]
First of all, thanks a lot! Did it actually resolve the time spent in that portion?

Let's see if we can do it more elegantly as a part of commit / commit_epilogue.

File: aptos-move/block-executor/src/txn_last_input_output.rs
Comment (sitalkedia) [reply aptos-move/block-executor/src/txn_last_input_output.rs:352]: We need to change the type here as IIUC, the output status might be empty for cases where we have error or skip rest. Am I missing something?

File: aptos-move/block-executor/src/executor.rs
Comment (sitalkedia) [reply aptos-move/block-executor/src/executor.rs:633]: Using a lock here is not expensive since in theory only one thread should access this lock. Atomic type might not be that clean as we have to initialize the atomic usize with -1 or `usize:max` instead of using `Option`. Let me know if you have something else in mind.

File: aptos-move/block-executor/src/txn_last_input_output.rs
Comment (gelash) [reply aptos-move/block-executor/src/txn_last_input_output.rs:352]: Oh, my bad, in my mind I was imagining doing it at commit (and then we shouldn't need the option, nor the index to reduce every time - we would just know how many got committed and fill it up)

File: aptos-move/block-executor/src/executor.rs
Comment (sitalkedia) [reply aptos-move/block-executor/src/executor.rs:612]: If it can be done cleanly in the refactor and that's something we can land soon, then I am fine abandoning this in favor of including it in the refactor. Otherwise, we can land this change and see if this can be done more nicely as a part of ongoing refactor? Thoughts?

File: aptos-move/block-executor/src/executor.rs
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:633]: yeah, with max, slightly cleaner maybe and just because it falls strictly within atomic semantics, but I agree this thing should be updated rarely, and even if not, wouldn't expect a performance hit.

File: aptos-move/block-executor/src/executor.rs
Comment (pr-aptos) [reply aptos-move/block-executor/src/executor.rs:612]: I would be in favor of landing this and trying it out. We can do a proper refactoring only if we see value.

Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:612]: I don't see an issue to refactor as needed after landing it, but I think we should land if we see value? We can check the effect already on the PR, right?

Comment (sitalkedia) [reply aptos-move/block-executor/src/executor.rs:612]: @gelash - Yes we are seeing around 4-5% TPS boost with executor benchmark.

Comment (pr-aptos) [reply aptos-move/block-executor/src/executor.rs:612]: Good point. I see we are already discussing the improvements in the slack thread.

Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:612]: so, once we have the numbers, happy to approve as is

Comment (sitalkedia) [reply aptos-move/block-executor/src/executor.rs:612]: Updated the test plan with the recent numbers.

**Review by gelash — APPROVED** [review]
happy to land then merge into commit_epilogue later.

File: aptos-move/block-executor/src/executor.rs
Comment (ziaptos) [reply aptos-move/block-executor/src/executor.rs:612]: I'm OK with landing this independently if we see value - tho further refactoring can bring more value.

File: aptos-move/block-executor/src/executor.rs
Comment (ziaptos) [reply aptos-move/block-executor/src/executor.rs:633]: I'd go with an atomic (relaxed read). Not that important for this PR [will refactor]

File: aptos-move/block-executor/src/executor.rs
Comment (ziaptos) [inline aptos-move/block-executor/src/executor.rs:641]: It's guaranteed that we do this once? right? maybe assert in debug mode?

**Review by ziaptos — COMMENTED** [review]
I'm ok with landing this, and continuing the progress.

I'd just make sure to document the exact setup and results that bring value. "10% on ~20k" is very vague. Let's have the exact machine used, exact set of parameters (git hash of the benchmarked version) - perhaps a link to the raw outputs of the measurements.

Would also be nice to separately benchmark the scalability of just this improvement.

**Comment by sitalkedia** [comment]
@gelash - Changed the logic to ensure the parallel result is exactly same as the sequential one. Please take another pass.

File: aptos-move/block-executor/src/executor.rs
Comment (sitalkedia) [reply aptos-move/block-executor/src/executor.rs]: I don't think this is guaranteed.. @gelash can clarify...

File: aptos-move/block-executor/src/executor.rs
Comment (sitalkedia) [reply aptos-move/block-executor/src/executor.rs]: My preference would be keeping the lock without atomic as IMO setting the atomic with max is unclean.. can look to change if any strong opinion.

**Comment by sitalkedia** [comment]
> I'm ok with landing this, and continuing the progress.
> 
> I'd just make sure to document the exact setup and results that bring value. "10% on ~20k" is very vague. Let's have the exact machine used, exact set of parameters (git hash of the benchmarked version) - perhaps a link to the raw outputs of the measurements.
> 
> Would also be nice to separately benchmark the scalability of just this improvement.

Updated the test plan with more information. Regarding separate benchmark for this improvement, I am not sure if that is more useful, we care about the e2e TPS than improvements in the micro-benchmarks..

**Comment by gelash** [comment]
> > I'm ok with landing this, and continuing the progress.
> > I'd just make sure to document the exact setup and results that bring value. "10% on ~20k" is very vague. Let's have the exact machine used, exact set of parameters (git hash of the benchmarked version) - perhaps a link to the raw outputs of the measurements.
> > Would also be nice to separately benchmark the scalability of just this improvement.
> 
> Updated the test plan with more information. Regarding separate benchmark for this improvement, I am not sure if that is more useful, we care about the e2e TPS than improvements in the micro-benchmarks..

We need a systematic approach on scalability, and there is no substitute for the micro benchmark for those purposes, in fact we should improve the microbenchmark, I think this is what @ziaptos meant as well, we should make sure we have a way to replicate and document the same load/setting/improvement there. But we do it progressively.

File: aptos-move/block-executor/src/executor.rs
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs]: yeah, unfortunately atm everything gets processed, and then this part made sure it was the same as sequential (i.e. first skip took effect), also likely more errors could appear. I even remember something like this, because we used to have parallel processing logic here before and then decided to simplify during the review process and keep the TODO

File: aptos-move/block-executor/src/executor.rs
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:650]: With the determinism issue, it does seem a lot to do it this way if indeed there is be a much cleaner (and more efficient, but not sure how noticeable) solution. And I feel more so after accepting and realizing I didn't think of smt I should have (and 99% even ran into the same thing long time ago).. So I am inclined to not rush it now, unless we really need this on the main asap, I'd rather us try it the other way? (coordinator goes sequentially and can stop the execution when it sees first abort/skiprest, we already do so for gas, workers can do the other part - and they are parallelized and pipelined with main work already, without spawning anew - unless I'm missing smt this seems to be what we need thanks to the rolling commit? - I don't think we had revisited this after rolling commit)

Comment (sitalkedia) [reply aptos-move/block-executor/src/executor.rs:650]: @gelash - Indeed the current approach is a bit convoluted to maintain the determinism. I am okay with waiting to land the block-stm refactor and doing it in a cleaner and potentially more efficient way. However, I would like to get this optimization in main in a week or so, so that we can include this in previewnet. Do you think that's a reasonable timeline? If not, I would be inclined to land the current change as a temporary approach, while the block-stm refactor is making progress.

File: aptos-move/block-executor/src/executor.rs
Comment (ziaptos) [reply aptos-move/block-executor/src/executor.rs]: Renaming the variable and using the vector size as the initialization value should be cleaner than a mutex IMO. OK with keeping the mutex tho.

File: aptos-move/block-executor/src/executor.rs
Comment (ziaptos) [inline aptos-move/block-executor/src/executor.rs:621]: The `min_error_index` mutex already guards the `maybe_err` - the second lock is unnecessary. Let's get around Rust's limits in a different way (use a tuple guarded by a mutex?).

**Comment by gelash** [comment]
> I'd like to see people agree on the next step before approve-ing the PR. Again I'm OK with both approaches.
> 
> Regarding the code tho - please address the inline comment about the double mutex usage.

If we don't land this we don't need to address changes. Let's figure out if the other approach works easily first?

**Pull request closed without merging**
